### ISSUE TITLE
Name missing

### ISSUE THREAD

**Issue by MartinMuzatko** [danielkrainas/node-netstat#22]
Hello!

I wanted to include `node-netstat` as opposed to [`netstat`](https://www.npmjs.com/package/netstat) because it gives me a better way of filtering and limiting. However I don't get any names of the programms.

This is what `netstat tulpn` gives me:

``` bash
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 192.168.99.10:53        0.0.0.0:*               LISTEN      306/named
tcp        0      0 127.0.0.1:53            0.0.0.0:*               LISTEN      306/named
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      -
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           364/avahi-daemon: r
udp        0      0 0.0.0.0:40475           0.0.0.0:*                           7618/dhcpd
udp        0      0 192.168.99.10:53        0.0.0.0:*                           306/named
udp        0      0 192.168.3.112:53        0.0.0.0:*                           306/named
udp        0      0 192.168.3.103:53        0.0.0.0:*                           306/named
udp        0      0 10.8.0.34:53            0.0.0.0:*                           306/named
udp        0      0 172.18.0.1:53           0.0.0.0:*                           306/named
udp        0      0 172.17.0.1:53           0.0.0.0:*                           306/named
udp        0      0 127.0.0.1:53            0.0.0.0:*                           306/named
udp        0      0 0.0.0.0:67              0.0.0.0:*                           7618/dhcpd
udp        0      0 0.0.0.0:68              0.0.0.0:*                           7154/dhclient
udp        0      0 0.0.0.0:68              0.0.0.0:*                           7011/dhclient
udp        0      0 0.0.0.0:68              0.0.0.0:*                           6043/dhclient

```

this is the console output of `node-netstat`

``` js
{ protocol: 'tcp',
  local: { port: 22, address: '192.168.3.103' },
  remote: { port: 49276, address: '192.168.3.109' },
  state: 'ESTABLISHED',
  pid: 0 }
{ protocol: 'tcp',
  local: { port: 41338, address: '192.168.3.103' },
  remote: { port: 443, address: '192.168.3.103' },
  state: 'TIME_WAIT',
  pid: 0 }
{ protocol: 'tcp',
  local: { port: 41300, address: '192.168.3.103' },
  remote: { port: 443, address: '192.168.3.103' },
  state: 'TIME_WAIT',
  pid: 0 }
{ protocol: 'tcp',
  local: { port: 52200, address: '192.168.3.103' },
  remote: { port: 80, address: '192.168.3.103' },
  state: 'TIME_WAIT',
  pid: 0 }
{ protocol: 'tcp',
  local: { port: 52310, address: '192.168.3.103' },
  remote: { port: 80, address: '192.168.3.103' },
  state: 'TIME_WAIT',
  pid: 0 }
```

Is there any way to include the name also and the other informations? Thank you in advance!

**Comment by danielkrainas** [comment]
Hi! My original use case that caused me to write node-netstat was to link apps to ports using PIDs but I can't remember what made it impractical. It may have had something to do with compatibility since PIDs aren't always available either. In any case, you are welcome to apply your own parser and commands to the library to use in place of the defaults and address your specific use case.

For an idea on how to write your parser, check out the current ones (really simple to write)

https://github.com/danielkrainas/node-netstat/blob/master/lib/parsers.js

To make use of your custom parser, you can override your target platform's parser and commands or create a custom platform key in `netstat.parsers` and `netstat.commands` to then specify when you actually execute `netstat`

```js
netstat.commands.mine = { cmd: 'netstat', args: ['-tulpn'] }
netstat.parsers.mine = customParserFunction
netstat({ platform: 'mine', ...})
```

Let me know if that clears things up. I'm happy to help further.

**Comment by MartinMuzatko** [comment]
`Hello. Yes that is clear to me.
I just hoped there would be existing tooling that helped me to avoid reinventing the wheel.

The reason why I was specifically looking for a netstat parser, was that I had trouble with bogus/unexpected output.
My original regex was `/\s+/` but that does not work in the case of `avahi-daemon` and sometimes `ssh`.

**Example:**
``` bash
# netstat -tupan // Guess I might just omit -a for the ESTABLISHED connection.
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      431/sshd           
tcp        0    272 192.168.1.242:22        192.168.1.129:44452     ESTABLISHED 1172/sshd: martin [
udp        0      0 0.0.0.0:48720           0.0.0.0:*                           451/avahi-daemon: r
```
**Source:** https://debian-handbook.info/browse/stable/sect.network-diagnosis-tools.html

I could also include in my regex to only go until `:` in strings with `/`, but I believe this will be not the end of the story ;)

Thanks for the help anyway. If you see it as option to extend the parser for linux to also include name by default, I'm happy to help!

From what I see, the name could be parsed correctly already: https://github.com/danielkrainas/node-netstat/blob/master/lib/parsers.js#L11
Maybe something is missing in the command? `-p` should be already enough, so maybe the name is omitted somewhere else? https://github.com/danielkrainas/node-netstat/blob/master/lib/netstat.js#L14

**Comment by danielkrainas** [comment]
@MartinMuzatko I remember now the hangup was with mac compatibility because netstat on mac doesn't output the pid or process name and you typically need to use a secondary tool like `lsof` to perform that resolution. It really isn't reinventing the wheel, in fact an expected use case for node-netstat. Here is a solution with a custom parser that adds the program name when available.

```
const netstat = require('node-netstat');
const { normalizeValues } = netstat.utils;

netstat.parsers.linux = function (line, callback) {
    var parts = line.split(/\s/).filter(String);
    if (!parts.length || parts.length < 7) {
        return;
    }

    var appName = null;
    var pid = parts.slice(6, parts.length).join(" ");
    if (pid.indexOf('/') > 0) {
      var pidParts = pid.split('/');
      pid = pidParts[0];
      appName = pidParts.slice(1, pidParts.length).join('/');
    }

    var item = {
        protocol: parts[0],
        local: parts[3],
        remote: parts[4],
        state: parts[5],
        pid: pid,
        appName: appName,
    };

    return callback(normalizeValues(item));
};

nn({}, item => console.log(item));
```

And that will output results with the `appName`

```
{ protocol: 'tcp',
  local: { port: 48120, address: '192.168.1.148' },
  remote: { port: 443, address: '198.252.206.25' },
  state: 'ESTABLISHED',
  pid: 6539,
  appName: 'chrome' }
{ protocol: 'tcp6',
  local: { port: 631, address: '::1' },
  remote: { port: NaN, address: null },
  state: 'LISTEN',
  pid: 0,
  appName: null }
```

**Comment by danielkrainas** [comment]
@MartinMuzatko I've published an update (v1.6.0) that includes an easier recipe for parsing `processName` on linux platform, [a code example is provided in the documentation.](https://github.com/danielkrainas/node-netstat#include-the-process-name-and-allow-filtering-on-it-linux-only). Hopefully this is a better solution for you.

**Comment by MartinMuzatko** [comment]
thank you a lot :)
